src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -272,26 +277,40 @@ string EncodeUrl (Uri url)
 				if(Logger.LogNet)
 					Logger.Log (LogLevel.Info, LOG_APP, " cancelled");
 
-				httpConnection.Disconnect ();
+				await DisconnectAsync (httpConnection).ConfigureAwait (continueOnCapturedContext: false);
 				cancellationToken.ThrowIfCancellationRequested ();
 			}
-			cancellationToken.Register (httpConnection.Disconnect);
 
-			if (httpConnection.DoOutput) {
-				using (var stream = await request.Content.ReadAsStreamAsync ()) {
-					await stream.CopyToAsync(httpConnection.OutputStream, 4096, cancellationToken)
-						.ConfigureAwait(false);
+			CancellationTokenRegistration cancelRegistration = default (CancellationTokenRegistration);
+			HttpStatusCode statusCode = HttpStatusCode.OK;
+			Uri connectionUri = null;
+
+			try {
+				cancelRegistration = cancellationToken.Register (() => {
+					DisconnectAsync (httpConnection).ContinueWith (t => {
+						if (t.Exception != null)
+							Logger.Log (LogLevel.Info, LOG_APP, $"Disconnection exception: {t.Exception}");
+					}, TaskScheduler.Default);
+				}, useSynchronizationContext: false);
+
+				if (httpConnection.DoOutput) {
+					using (var stream = await request.Content.ReadAsStreamAsync ()) {
+						await stream.CopyToAsync(httpConnection.OutputStream, 4096, cancellationToken)
+							.ConfigureAwait(false);
+					}
 				}
+
+				statusCode = await Task.Run (() => (HttpStatusCode)httpConnection.ResponseCode).ConfigureAwait (false);
+				connectionUri = new Uri (httpConnection.URL.ToString ());
+			} finally {
+				cancelRegistration.Dispose ();
 			}
 
 			if (cancellationToken.IsCancellationRequested) {
-				httpConnection.Disconnect();
+				await DisconnectAsync (httpConnection).ConfigureAwait (continueOnCapturedContext: false);
```

Comment:
@alanmcgovern: please correct me if I'm wrong, but I thought that prior discussion was that it was pointless to do `.ConfigureAwait (continueOnCapturedContext: false)` in a method after having already called it. Here, `.ConfigureAwait()` is called 7 and 11 lines above, so is this invocation necessary?

I'm still not clear on all these async rules & conventions...